backend/cfg/cfg_dataflow.ml

```diff
@@ -1,151 +1,381 @@
-    let value = Option.value init ~default:D.top in
-    (* The need to have several blocks in the initial work set stems from the
-       fact that we currently need to consider all trap handlers as alive. *)
+    Cfg.iter_blocks cfg ~f:(fun label _block ->
+        if not (Label.Tbl.mem mapping label)
+        then
+          let (_ : priority_helper) = strong_connect label in
+          assert (Stack.is_empty stack));
+    assert (Label.Tbl.length priorities = Label.Tbl.length cfg.blocks);
+    priorities
+
+  let update_instr : work_state -> int -> instr_domain -> unit =
+   fun t instr_id value ->
+    match t.map_instr with
+    | None -> ()
+    | Some map_instr -> Instr.Tbl.replace map_instr instr_id value
+
+  let create :
+      Cfg.t ->
+      init:(Cfg.basic_block -> Transfer_domain.t option) ->
+      store_instr:bool ->
+      work_state =
+   fun cfg ~init ~store_instr ->
+    let priorities = compute_priorities cfg in
+    let queue = WorkSet.create ~priorities in
+    let map_block = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let map_instr =
+      if store_instr
+      then
+        let map_instr =
+          Instr.Tbl.create (Label.Tbl.length cfg.Cfg.blocks * 16)
+        in
+        Some map_instr
+      else None
+    in
+    let t = { cfg; queue; map_block; map_instr } in
     Cfg.iter_blocks cfg ~f:(fun label block ->
-        if Label.equal label cfg.entry_label || block.is_trap_handler
-        then set := WorkSet.add { WorkSetElement.label; value } !set);
-    map, set
+        match init block with
+        | None -> ()
+        | Some value ->
+          Label.Tbl.replace map_block label value;
+          WorkSet.add t.queue label);
+    t
 
-  let remove_and_return :
-      Cfg.t -> WorkSet.t ref -> WorkSetElement.t * Cfg.basic_block =
-   fun cfg set ->
-    let element = WorkSet.choose !set in
-    set := WorkSet.remove element !set;
-    element, Cfg.get_block_exn cfg element.label
+  let get_res_block t = t.map_block
 
-  let run :
-      Cfg.t -> ?max_iteration:int -> ?init:domain -> unit -> (map, map) Result.t
-      =
-   fun cfg ?(max_iteration = max_int) ?init () ->
-    let res, work_set = create cfg ~init in
+  let get_res_instr_exn t = Option.get t.map_instr
+
+  let run ~max_iteration work_state =
     let iteration = ref 0 in
-    while (not (WorkSet.is_empty !work_set)) && !iteration < max_iteration do
+    (* note: `handler_map` contains the value at the *start* of the block. *)
+    while
+      (not (WorkSet.is_empty work_state.queue)) && !iteration < max_iteration
+    do
       incr iteration;
-      let element, block = remove_and_return cfg work_set in
-      let ({ normal; exceptional } : T.t) =
-        transfer_block element.value block
+      let element = WorkSet.remove_and_return work_state.queue in
+      let current_block = Cfg.get_block_exn work_state.cfg element in
+      let current_value =
+        match Label.Tbl.find work_state.map_block current_block.start with
```

Comment:
Would it be beneficial to populate
the table at the top to get rid of
the pattern matching? (I am fine
either way)

(I think in an earlier version I
wanted to be able to distinguish
between not-visited and visited
with bottom, but I am not sure
that really makes sense.)

Reply:
Done,